### PR TITLE
Perform ExistentialSpecializer when SoleConformingType is known

### PR DESCRIPTION
**Pull request by rajbarik**
The criteria for ExistentialSpecializer is now based on  an argument having "findInitExistential or SoleConformingType".

### CODE REVIEW

File: lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp
Comment (rajbarik) [inline lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp]: This method checks if a SoleConformingType can be determined for the argument. Some parts of code (5-6 lines) overlap with what was done in SILCombiner, but there are lot of disparities to make it a new APi in Existential.h altogether.

File: lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp
Comment (rajbarik) [inline lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp]: Added to filter out inout args, which otherwise would be fed to openExistentialRef as non-objects!

File: test/SILOptimizer/existential_transform_soletype.swift
Comment (rajbarik) [inline test/SILOptimizer/existential_transform_soletype.swift]: opt1 does not get optimized since the caller foo is under @_optimize(none).

**Comment by rajbarik** [comment]
@slavapestov @atrick Can you review this small PR? This one extends ExistentialSpecializer to also perform the optimization when SoleConformingType can be found.

File: lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp
Comment (rajbarik) [reply lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp]: @atrick You had a similar concern from SILCombiner, but the overlap is minor to warrant a new API?

File: lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp
Comment (rajbarik) [inline lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp]: @atrick fyi

File: lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp
Comment (atrick) [inline lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp]: This should be a standalone API because it has special logic to reason about possible conformances. That logic should only exist in one place in the compiler.

File: test/SILOptimizer/existential_transform_soletype.swift
Comment (atrick) [reply test/SILOptimizer/existential_transform_soletype.swift]: It's generally better to use `.sil` test cases for individual passes, but not an absolute requirement. The source is definitely helpful, but that can be included as a comment.

**Review by atrick — APPROVED** [review]
Otherwise, LGTM.

File: lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp
Comment (rajbarik) [reply lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp]: Originally I had this in findSoleConformingType function of ProtocolConformanceAnalysis, however since this also required ClassHierarchyAnalysis, we had split. I can put this in Existential.cpp in Utils. This would be a wrapper for findSoleConformingType.

Comment (atrick) [reply lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp]: I'm just trying to avoid duplicating logic that reasons about the type system in multiple places in the SILOptimizer. As an alternative, we could just have a `CHA->mayHaveSubclasses` query.

**Comment by rajbarik** [comment]
@atrick after you merge your changes https://github.com/apple/swift/pull/21316, we should clean this up and get this merged with cleaner interface.

**Comment by atrick** [comment]
@rajbarik I finished rebasing 
https://github.com/apple/swift/pull/21316 but just missed the branch lockdown. Of course, you could check this branch out yourself from my repo if you want to cleanup this PR. I'll merge my PR without further changes on Jan 2.

**Comment by rajbarik** [comment]
@atrick please check. I extracted out the common code between SILCombiner and ExistentialSpecializer to ProtocolConformanceAnalysis. Have a look.

**Comment by rajbarik** [comment]
@atrick, fyi, this needs your approval. thanks.

**Comment by atrick** [comment]
@rajbarik it looks like I already approved this, but I think you need to add a commit message.

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by atrick** [comment]
@swift-ci test.

**Comment by slavapestov** [comment]
@rajbarik Are you able to build and test Swift on Linux?

```
18:39:35 The following tests FAILED:
18:39:35 	  1 - TestFoundation (SEGFAULT)
```

**Comment by rajbarik** [comment]
Update: Found a linux box and installed Swift on it. It built fine. When I run, I get a lot of precompiled header errors, nothing like Foundation SIGFAULT. 
@atrick any clue how to go about handling the Linux error that Slava is pointing out?

**Comment by rajbarik** [comment]
@atrick //cc @airspeedswift @eeckstein could you please look at this one when you get time? I am unable to reproduce this on a linux box at my end. I am in the process of writing a benchmark for https://github.com/apple/swift/pull/19820 -- this one needs to be committed before I commit the benchmark.

**Comment by atrick** [comment]
@swift-ci test linux platform.

**Comment by atrick** [comment]
I don't think I'll have time to debug this week. Let's get the results again. Maybe it's unrelated to this PR.

**Pull request merged**
